File: src/commands/embed.rs
use std::env;
use std::fs;
use std::io::{self, IsTerminal, Read};

use clap::{Args, ValueEnum};
use serde::Serialize;

use crate::config::{self, ProfileConfig};
use crate::rchain::embeddings::{
    self, ChunkStrategy, EmbeddingProvider, EmbeddingResult, EmbeddingsConfig,
};

#[derive(Debug, Args, Clone)]
pub struct EmbedArgs {
    #[arg(long)]
    pub profile: Option<String>,

    #[arg(long, value_enum)]
    pub provider: Option<ProviderArg>,

    #[arg(long)]
    pub model: Option<String>,

    #[arg(long)]
    pub chunk_size: Option<usize>,

    #[arg(long)]
    pub chunk_overlap: Option<usize>,

    #[arg(long, value_enum)]
    pub chunk_strategy: Option<ChunkStrategyArg>,

    #[arg(long, value_enum)]
    pub output: Option<OutputFormatArg>,

    #[arg(long)]
    pub json: bool,

    #[arg(long)]
    pub file: Option<std::path::PathBuf>,

    input: Option<String>,
}

#[derive(Debug, Clone, Copy, ValueEnum)]
pub enum ProviderArg {
    Openai,
    Fireworks,
}

#[derive(Debug, Clone, Copy, ValueEnum)]
pub enum ChunkStrategyArg {
    Paragraph,
    Sentence,
    Token,
}

#[derive(Debug, Clone, Copy, ValueEnum)]
pub enum OutputFormatArg {
    Text,
    Json,
}

impl From<ChunkStrategyArg> for ChunkStrategy {
    fn from(arg: ChunkStrategyArg) -> Self {
        match arg {
            ChunkStrategyArg::Paragraph => ChunkStrategy::Paragraph,
            ChunkStrategyArg::Sentence => ChunkStrategy::Sentence,
            ChunkStrategyArg::Token => ChunkStrategy::Token,
        }
    }
}

impl From<ProviderArg> for EmbeddingProvider {
    fn from(arg: ProviderArg) -> Self {
        match arg {
            ProviderArg::Openai => EmbeddingProvider::Openai,
            ProviderArg::Fireworks => EmbeddingProvider::Fireworks,
        }
    }
}

#[derive(Debug, Serialize)]
struct JsonOutput {
    provider: String,
    model: String,
    chunks: Vec<String>,
    embeddings: Vec<Vec<f64>>,
}

pub fn run(cli: EmbedArgs) -> Result<(), String> {
    let profile = resolve_profile(cli.profile.as_deref())?;
    let provider = resolve_provider(cli.provider, &profile)?;
    let model = resolve_model(cli.model, &profile)?;
    let chunk_size = resolve_chunk_size(cli.chunk_size, &profile)?;
    let chunk_overlap = resolve_chunk_overlap(cli.chunk_overlap, &profile)?;
    let chunk_strategy = resolve_chunk_strategy(cli.chunk_strategy, &profile)?;
    let output_format = resolve_output_format(cli.output, cli.json)?;

    let input_text = resolve_input(cli.input, cli.file)?;

    let config = EmbeddingsConfig {
        provider,
        model,
        chunk_size,
        chunk_overlap,
        chunk_strategy,
    };

    let result = embeddings::embed_texts(&config, &[input_text]).map_err(|err| err.to_string())?;

    render_output(&result, output_format)?;

    Ok(())
}

fn resolve_profile(profile_name: Option<&str>) -> Result<ProfileConfig, String> {
    match profile_name {
        Some(name) => config::load_profile(name),
        None => Ok(ProfileConfig::default()),
    }
}

fn resolve_provider(
    cli_provider: Option<ProviderArg>,
    profile: &ProfileConfig,
) -> Result<EmbeddingProvider, String> {
    if let Some(provider) = cli_provider {
        return Ok(provider.into());
    }

    if let Ok(raw) = env::var("MP_PROVIDER") {
        return parse_provider_value(&raw, "MP_PROVIDER");
    }

    if let Some(provider) = &profile.provider {
        return parse_provider_value(provider, "profile provider");
    }

    Ok(EmbeddingProvider::Fireworks)
}

fn parse_provider_value(raw: &str, source: &str) -> Result<EmbeddingProvider, String> {
    match raw.trim().to_ascii_lowercase().as_str() {
        "openai" => Ok(EmbeddingProvider::Openai),
        "fireworks" => Ok(EmbeddingProvider::Fireworks),
        other => Err(format!(
            "Invalid {source} '{other}'. Supported values: openai, fireworks."
        )),
    }
}

fn resolve_model(cli_model: Option<String>, profile: &ProfileConfig) -> Result<String, String> {
    if let Some(model) = cli_model {
        let trimmed = model.trim();
        if !trimmed.is_empty() {
            return Ok(trimmed.to_string());
        }
    }

    if let Ok(model) = env::var("MP_MODEL") {
        let trimmed = model.trim();
        if !trimmed.is_empty() {
            return Ok(trimmed.to_string());
        }
    }

    if let Some(model) = &profile.embedding_model {
        let trimmed = model.trim();
        if !trimmed.is_empty() {
            return Ok(trimmed.to_string());
        }
    }

    Ok(EmbeddingProvider::Fireworks.as_str().to_string())
}

fn resolve_chunk_size(cli_size: Option<usize>, profile: &ProfileConfig) -> Result<usize, String> {
    if let Some(size) = cli_size {
        if size == 0 {
            return Err("Chunk size must be greater than 0.".to_string());
        }
        return Ok(size);
    }

    if let Some(size) = profile.chunk_size {
        if size == 0 {
            return Err("Chunk size in profile must be greater than 0.".to_string());
        }
        return Ok(size);
    }

    Ok(8000)
}

fn resolve_chunk_overlap(
    cli_overlap: Option<usize>,
    profile: &ProfileConfig,
) -> Result<usize, String> {
    if let Some(overlap) = cli_overlap {
        if overlap > 100 {
            return Err("Chunk overlap must be between 0 and 100.".to_string());
        }
        return Ok(overlap);
    }

    if let Some(overlap) = profile.chunk_overlap {
        if overlap > 100 {
            return Err("Chunk overlap in profile must be between 0 and 100.".to_string());
        }
        return Ok(overlap);
    }

    Ok(10)
}

fn resolve_chunk_strategy(
    cli_strategy: Option<ChunkStrategyArg>,
    profile: &ProfileConfig,
) -> Result<ChunkStrategy, String> {
    if let Some(strategy) = cli_strategy {
        return Ok(strategy.into());
    }

    if let Some(raw) = &profile.chunk_strategy {
        if let Some(strategy) = ChunkStrategy::from_str(raw) {
            return Ok(strategy);
        }
        return Err(format!(
            "Invalid chunk strategy '{raw}'. Supported values: paragraph, sentence, token."
        ));
    }

    Ok(ChunkStrategy::Paragraph)
}

fn resolve_output_format(
    output: Option<OutputFormatArg>,
    json: bool,
) -> Result<OutputFormatArg, String> {
    if json {
        return Ok(OutputFormatArg::Json);
    }

    if let Some(output) = output {
        return Ok(output);
    }

    Ok(OutputFormatArg::Text)
}

fn resolve_input(
    cli_input: Option<String>,
    file: Option<std::path::PathBuf>,
) -> Result<String, String> {
    if let Some(text) = cli_input {
        let trimmed = text.trim();
        if !trimmed.is_empty() {
            return Ok(trimmed.to_string());
        }
    }

    if let Some(path) = file {
        return fs::read_to_string(&path)
            .map_err(|err| format!("Failed to read file '{}': {}", path.display(), err));
    }

    if io::stdin().is_terminal() {
        return Err("No input provided. Pass an argument, --file, or pipe stdin.".to_string());
    }

    let mut buffer = String::new();
    io::stdin()
        .read_to_string(&mut buffer)
        .map_err(|err| format!("Failed to read stdin: {err}"))?;

    let text = buffer.trim().to_string();
    if text.is_empty() {
        return Err("Input is empty.".to_string());
    }

    Ok(text)
}

fn render_output(result: &EmbeddingResult, format: OutputFormatArg) -> Result<(), String> {
    match format {
        OutputFormatArg::Text => render_text(result),
        OutputFormatArg::Json => render_json(result),
    }
}

fn render_text(result: &EmbeddingResult) -> Result<(), String> {
    for embedding in &result.embeddings {
        let line = embedding
            .iter()
            .map(|v| v.to_string())
            .collect::<Vec<_>>()
            .join(",");
        println!("{line}");
    }
    Ok(())
}

fn render_json(result: &EmbeddingResult) -> Result<(), String> {
    let output = JsonOutput {
        provider: result.provider.clone(),
        model: result.model.clone(),
        chunks: result.chunks.clone(),
        embeddings: result.embeddings.clone(),
    };

    let json =
        serde_json::to_string(&output).map_err(|err| format!("Failed to serialize JSON: {err}"))?;

    println!("{json}");
    Ok(())
}

File: mpipe/commands/embed.py
# SPDX-License-Identifier: GPL-3.0-or-later
#
# mpipe - Multi-provider LLM CLI tools
# Copyright (C) 2026 Nicolas Schaetti and contributors
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

"""mpipe.commands.embed module.

Notes
-----
This module is part of the Python port of the `mpipe` project.
"""
from __future__ import annotations

import sys
import os
from pathlib import Path

import click

from mpipe.config import ProfileConfig, load_profile
from mpipe.console import console, print_json
from mpipe.rchain import embeddings
from mpipe.rchain.embeddings import ChunkStrategy, EmbeddingProvider, EmbeddingsConfig


@click.command("embed")
@click.option("--profile")
@click.option("--provider", type=click.Choice(["openai", "fireworks"]))
@click.option("--model")
@click.option("--chunk-size", type=int)
@click.option("--chunk-overlap", type=int)
@click.option("--chunk-strategy", type=click.Choice(["paragraph", "sentence", "token"]))
@click.option("--output", type=click.Choice(["text", "json"]))
@click.option("--json", "json_output", is_flag=True)
@click.option("--file", "file_path", type=click.Path(path_type=Path))
@click.argument("input_text", required=False)
def embed_command(
    profile: str | None,
    provider: str | None,
    model: str | None,
    chunk_size: int | None,
    chunk_overlap: int | None,
    chunk_strategy: str | None,
    output: str | None,
    json_output: bool,
    file_path: Path | None,
    input_text: str | None,
) -> None:
    """Embed command.

    Parameters
    ----------
    profile : str | None
        Argument value.
    provider : str | None
        Argument value.
    model : str | None
        Argument value.
    chunk_size : int | None
        Argument value.
    chunk_overlap : int | None
        Argument value.
    chunk_strategy : str | None
        Argument value.
    output : str | None
        Argument value.
    json_output : bool
        Argument value.
    file_path : Path | None
        Argument value.
    input_text : str | None
        Argument value.

    Returns
    -------
    None
        Returned value.
    """
    try:
        profile_cfg = load_profile(profile) if profile else ProfileConfig()
        emb_provider = _resolve_provider(provider, profile_cfg)
        emb_model = _resolve_model(model, profile_cfg)
        emb_chunk_size = _resolve_chunk_size(chunk_size, profile_cfg)
        emb_chunk_overlap = _resolve_chunk_overlap(chunk_overlap, profile_cfg)
        emb_chunk_strategy = _resolve_chunk_strategy(chunk_strategy, profile_cfg)
        output_format = "json" if json_output else (output or "text")
        text = _resolve_input(input_text, file_path)

        config = EmbeddingsConfig(
            provider=emb_provider,
            model=emb_model,
            chunk_size=emb_chunk_size,
            chunk_overlap=emb_chunk_overlap,
            chunk_strategy=emb_chunk_strategy,
        )
        result = embeddings.embed_texts(config, [text])

        if output_format == "json":
            print_json(
                {
                    "provider": result.provider,
                    "model": result.model,
                    "chunks": result.chunks,
                    "embeddings": result.embeddings,
                }
            )
            return
        # end if

        for vector in result.embeddings:
            console.print(",".join(str(value) for value in vector))
        # end for
    except Exception as err:
        raise click.ClickException(str(err)) from err
    # end try
# end def embed_command


def _resolve_provider(cli_provider: str | None, profile: ProfileConfig) -> EmbeddingProvider:
    """ resolve provider.

    Parameters
    ----------
    cli_provider : str | None
        Argument value.
    profile : ProfileConfig
        Argument value.

    Returns
    -------
    EmbeddingProvider
        Returned value.
    """
    raw = cli_provider or (profile.provider.strip() if profile.provider else "fireworks")
    return EmbeddingProvider(raw)
# end def _resolve_provider


def _resolve_model(cli_model: str | None, profile: ProfileConfig) -> str:
    """ resolve model.

    Parameters
    ----------
    cli_model : str | None
        Argument value.
    profile : ProfileConfig
        Argument value.

    Returns
    -------
    str
        Returned value.
    """
    if cli_model and cli_model.strip():
        return cli_model.strip()
    # end if
    if profile.embedding_model and profile.embedding_model.strip():
        return profile.embedding_model.strip()
    # end if
    env_model = os.getenv("MP_MODEL")
    if env_model and env_model.strip():
        return env_model.strip()
    # end if
    return EmbeddingProvider.FIREWORKS.as_str()
# end def _resolve_model


def _resolve_chunk_size(cli_chunk_size: int | None, profile: ProfileConfig) -> int:
    """ resolve chunk size.

    Parameters
    ----------
    cli_chunk_size : int | None
        Argument value.
    profile : ProfileConfig
        Argument value.

    Returns
    -------
    int
        Returned value.
    """
    value = cli_chunk_size if cli_chunk_size is not None else profile.chunk_size
    if value is None:
        return 8000
    # end if
    if value <= 0:
        raise ValueError("Chunk size must be greater than 0.")
    # end if
    return value
# end def _resolve_chunk_size


def _resolve_chunk_overlap(cli_chunk_overlap: int | None, profile: ProfileConfig) -> int:
    """ resolve chunk overlap.

    Parameters
    ----------
    cli_chunk_overlap : int | None
        Argument value.
    profile : ProfileConfig
        Argument value.

    Returns
    -------
    int
        Returned value.
    """
    value = cli_chunk_overlap if cli_chunk_overlap is not None else profile.chunk_overlap
    if value is None:
        return 10
    # end if
    if value < 0 or value > 100:
        raise ValueError("Chunk overlap must be between 0 and 100.")
    # end if
    return value
# end def _resolve_chunk_overlap


def _resolve_chunk_strategy(cli_chunk_strategy: str | None, profile: ProfileConfig) -> ChunkStrategy:
    """ resolve chunk strategy.

    Parameters
    ----------
    cli_chunk_strategy : str | None
        Argument value.
    profile : ProfileConfig
        Argument value.

    Returns
    -------
    ChunkStrategy
        Returned value.
    """
    raw = cli_chunk_strategy or profile.chunk_strategy or "paragraph"
    strategy = ChunkStrategy.from_str(raw)
    if strategy is None:
        raise ValueError(f"Invalid chunk strategy '{raw}'. Supported values: paragraph, sentence, token.")
    # end if
    return strategy
# end def _resolve_chunk_strategy


def _resolve_input(cli_input: str | None, file_path: Path | None) -> str:
    """ resolve input.

    Parameters
    ----------
    cli_input : str | None
        Argument value.
    file_path : Path | None
        Argument value.

    Returns
    -------
    str
        Returned value.
    """
    if cli_input and cli_input.strip():
        return cli_input.strip()
    # end if
    if file_path is not None:
        return file_path.read_text(encoding="utf-8")
    # end if
    if sys.stdin.isatty():
        raise ValueError("No input provided. Pass an argument, --file, or pipe stdin.")
    # end if
    data = sys.stdin.read().strip()
    if not data:
        raise ValueError("Input is empty.")
    # end if
    return data
# end def _resolve_input
